client/man/sxreport-client.1.in
.TH SXREPORT-CLIENT "1" "September 2014" "sxreport-client @VERSION@" "Skylable SX Manual"
.SH NAME
sxreport-client \- generate a report about the system and client configuration
.SH SYNOPSIS
.B sxreport-client
[\fI\,OPTIONS\/\fR]
.SH DESCRIPTION
This tool automatically generates a report about the system, client configuration, libraries and various dependencies, which can be useful when reporting a bug.
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-\-full\-help\fR
Print help, including hidden options, and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,PATH\/\fR
Save output to a given file. By default the report will be saved in the current directory as sxreport\-client\-<timestamp>.log.
.TP
\fB\-c\fR, \fB\-\-config\-dir\fR=\fI\,PATH\/\fR
Path to the SX configuration directory (default: ~/.sx)
.TP
\fB\-f\fR, \fB\-\-filter\-dir\fR=\fI\,PATH\/\fR
Path to the SX filter directory (default: @SX_FILTER_DIR@)
.SH SEE ALSO
\fBsxinit\fR(1)


NEWS
1.0
---

This is the first stable release of Skylable SX. The changes
from 0.9 include:

  * New user manual covering all important cluster and
    client operations (see doc/manual/manual.pdf or 
    http://www.skylable.com/products/sx/manual)
  * Support for replacing and rebuilding of broken nodes (see
    the manual for instructions on cluster healing)
  * New tool 'sxmv' for moving and renaming files (see 'man sxmv')
  * 'sxcp --ignore-errors' (ignore errors during recursive operations
    and report them at the end)
  * 'sxcp --exclude/--include' (skip or process files matching specific
    patterns)
  * 'sxvol modify' (change configuration of existing volumes)
  * 'sxadm --resize' (proportionally grow or shrink the entire cluster)
  * 'sxadm node --check' improvements (better detect local problems,
    eg. with faulty hard drives)
  * 'sxadm node --extract' (recover local storage in emergency situations)
  * Many improvements and bugfixes

0.9 (Final Beta)
----------------

This is a final beta of Skylable SX! Changes in this release include:

  * New garbage collector implementation, both more robust and scalable
  * Full support for file revisions (see 'man sxrev' and 'man sxvol' for
    details on how to manage and work with the revisions)
  * "undelete" filter (see 'sxvol filter --info undelete')
  * Usability improvements, including new man pages for the client tools
  * Many minor improvements and bugfixes

0.4 (Fourth Beta)
-----------------

This release contains important enhancements and bugfixes, including:

  * Automatic rebalance when nodes are added, modified or removed
    from the cluster
  * aes256 filter now uses the AES-256-CBC-HMAC-512 mode and bcrypt for
    key derivation
  * sxcp --one-file-system support (to not cross filesystem boundaries)
  * sxcp can now limit the number of connections (global and per host), see
    --full-help for usage information
  * Many usability enhancements in the client tools
  * Other improvements and bugfixes

0.3 (Third Beta)
----------------

The key features and improvements in this release include:

  * Support for using a custom port number (configure with sxsetup --advanced).
    This also allows sxsetup to run without root privileges when SX is
    configured to run on a port higher than 1024.
  * sxcp now displays a progress bar while transferring files
  * sxcp --bwlimit support (to set a bandwidth usage limit)
  * configure --with-system-libs support (to force the use of system libraries
    instead of those bundled in 3rdparty/)
  * Many minor improvements and bugfixes

0.2 (Second Beta)
-----------------

Changes in this release include:

  * Support for building with either NSS or OpenSSL
  * Minimum Curl version supported is now 7.34.0 (configure will use the
    embedded version if the system's one is not supported - this allows
    out-of-box build on CentOS 6)
  * aes256 filter now generates per-volume salts; in order to preserve
    deduplication support across different volumes using the same password
    you can force a static salt, eg. --filter aes256="salt:af6a812ad0b1cb22"
  * sxinit --force-reinit support (to recreate the entire local cluster config)
  * sxinit --list support (to list all configured clusters and profiles)
  * sxinit --alias support (to assign an alias for easy access), see
    https://wiki.skylable.com/wiki/Sxinit#Aliases
  * sxls -l now also lists filters used by volumes
  * Other bugfixes and improvements

0.1 (First Beta)
----------------

Initial public release of Skylable SX.


client/src/tools/sxreport-client/sxreport-client.c
/*
 *  Copyright (C) 2012-2014 Skylable Ltd. <info-copyright@skylable.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2 as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 *  Special exception for linking this software with OpenSSL:
 *
 *  In addition, as a special exception, Skylable Ltd. gives permission to
 *  link the code of this program with the OpenSSL library and distribute
 *  linked combinations including the two. You must obey the GNU General
 *  Public License in all respects for all of the code used other than
 *  OpenSSL. You may extend this exception to your version of the program,
 *  but you are not obligated to do so. If you do not wish to do so, delete
 *  this exception statement from your version.
 */

#include "default.h"
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <limits.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "cmdline.h"

#include "sx.h"
#include "../../../../libsx/src/sxreport.h"
#include "../../../../libsx/src/misc.h"
#include "version.h"

static int filter_list(sxc_client_t *sx)
{
    const sxf_handle_t *filters;
    int count, i;

    filters = sxc_filter_list(sx, &count);
    if(!filters) {
	sxi_notice(sx, "No filters available\n");
	return 1;
    }
    for(i = 0; i < count; i++) {
        const sxc_filter_t *f = sxc_get_filter(&filters[i]);
        sxi_info(sx, "'%s' filter details:", f->shortname);
        sxi_info(sx,"\tShort description: %s", f->shortdesc);
        sxi_info(sx,"\tSummary: %s", f->summary);
        sxi_info(sx,"\tOptions: %s", f->options ? f->options : "No options");
        sxi_info(sx,"\tUUID: %s", f->uuid);
        sxi_info(sx,"\tType: %s", f->tname);
        sxi_info(sx,"\tVersion: %d.%d", f->version[0], f->version[1]);
    }

    return 0;
}

static const char *get_filter_dir(sxc_client_t *sx, const char *fdir)
{
    const char *pt;
    if(fdir)
        return fdir;
    pt = sxi_getenv("SX_FILTER_DIR");
    if(pt)
        return pt;
    return SX_FILTER_DIR;
}

int main(int argc, char **argv) {
    sxc_client_t *sx;
    sxc_logger_t log;
    struct gengetopt_args_info args;
    char file[1024];
    const char *filter_dir;

    if(cmdline_parser(argc, argv, &args))
	return 1;

    if(args.version_given) {
	printf("%s %s\n", CMDLINE_PARSER_PACKAGE, SRC_VERSION);
	cmdline_parser_free(&args);
	return 0;
    }

    if(args.output_given)
        snprintf(file, sizeof(file), "%s", args.output_arg);
    else
	snprintf(file, sizeof(file), "sxreport-client-%ld.log", (long)time(NULL));

    umask(077);
    if(!(sx = sxc_init(SRC_VERSION, sxc_file_logger(&log, argv[0], file, 1), NULL, NULL))) {
	cmdline_parser_free(&args);
	return 1;
    }
    sxc_set_confdir(sx, args.config_dir_arg);
    sxc_set_verbose(sx, 1);

    filter_dir = get_filter_dir(sx, args.filter_dir_arg);
    sxc_filter_loadall(sx, filter_dir);
    sxi_report_build_info(sx);
#define INFO_PKGCONFIG ""
#define INFO_PKGCONFIG_LIBDIR ""
#define INFO_PKGCONFIG_PATH ""
    sxi_report_build_flags(sx);
    sxi_report_library_versions(sx, SRC_VERSION);
    sxi_report_system_info(sx);
    sxi_report_limits(sx);
    sxi_report_section(sx, "Filters");
    sxi_info(sx,"Default filter directory: %s\n", SX_FILTER_DIR);
    sxi_info(sx,"Current filter directory: %s\n", filter_dir);
    filter_list(sx);
    sxi_report_configuration(sx, args.config_dir_arg);
    printf("Report stored in %s\n", file);
    printf("You can attach it to a bugreport at %s\n", PACKAGE_BUGREPORT);
    cmdline_parser_free(&args);
    sxc_shutdown(sx, 0);
    return 0;
}
